base/event.jl

```diff
@@ -199,45 +199,53 @@ function throwto(t::Task, exc::ANY)
     return yieldto(t)
 end
 
-function ensure_self_descheduled()
-    # return a queued task to the runnable state
+function ensure_rescheduled(othertask::Task)
     ct = current_task()
+    if ct !== othertask && othertask.state == :runnable
+        # we failed to yield to othertask
+        # return it to the head of the queue to be scheduled later
+        unshift!(Workqueue, othertask)
+        othertask.state = :queued
+    end
     if ct.state == :queued
+        # if the current task was queued,
+        # also need to return it to the runnable state
+        # before throwing an error
         i = findfirst(Workqueue, ct)
         i == 0 || deleteat!(Workqueue, i)
         ct.state = :runnable
     end
     nothing
 end
 
+@noinline function poptask()
+    t = shift!(Workqueue)
+    if t.state != :queued
+        # assume this somehow got queued twice,
+        # probably broken now, but try discarding this switch and keep going
+        # can't throw here, because it's probably not the fault of the caller to wait
+        # and don't want to use print() here, because that may try to incur a task switch
+        ccall(:jl_safe_printf, Void, (Ptr{UInt8}, Int32...),
+            "\nWARNING: Workqueue inconsistency detected: shift!(Workqueue).state != :queued\n")
+        return
+    end
+    t.state = :runnable
+    return Ref(t)
+end
+
 function wait()
     while true
         if isempty(Workqueue)
             c = process_events(true)
-            if c==0 && eventloop()!=C_NULL && isempty(Workqueue)
+            if c == 0 && eventloop() != C_NULL && isempty(Workqueue)
                 # if there are no active handles and no runnable tasks, just
                 # wait for signals.
                 pause()
             end
         else
-            let t = shift!(Workqueue)
-                if t.state != :queued
-                    # assume this somehow got queued twice,
-                    # probably broken now, but try discarding this switch and keep going
-                    # can't throw here, because it's probably not the fault of the caller to wait
-                    # and don't want to use print() here, because that may try to incur a task switch
-                    ccall(:jl_safe_printf, Void, (Ptr{UInt8}, Int32...),
-                        "\nWARNING: Workqueue inconsistency detected: shift!(Workqueue).state != :queued\n")
```

Comment:
It's just moved. I needed it to be a separate function to ensure that the gc-root for `t` gets dropped and that this function is just left with `Ref(t)`.